File: BinarySearch/Leetcode/Medium/PeakIn2D/PeakIn2D.cpp
// https://leetcode.com/problems/find-a-peak-element-ii/
#include<bits/stdc++.h>
using namespace std;
int maxInColumn(vector<vector<int>> &mat, int j)
{
    int maxR = 0;
    for (int i = 1; i < mat.size(); i++)
    {
        if (mat[maxR][j] < mat[i][j])
        {
            maxR = i;
        }
    }
    return maxR;
}

vector<int> findPeakGrid(vector<vector<int>> &mat)
{
    int n = mat[0].size();
    int l = 0, u = n - 1;
    while (l <= u)
    {
        int mid = u - (u - l) / 2;
        int maxR = maxInColumn(mat, mid);
        int left = mid == 0 ? -1 : mat[maxR][mid - 1];
        int right = mid == n - 1 ? -1 : mat[maxR][mid + 1];
        if (mat[maxR][mid] > left && mat[maxR][mid] > right)
        {
            return {maxR, mid};
        }
        else if (mat[maxR][mid] > left)
        {
            l = mid + 1;
        }
        else
        {
            u = mid - 1;
        }
    }
    return {-1, -1};
}
int main(void) {
    
    return 0;
}

File: BinarySearch/Leetcode/Medium/PeakIn2D/PeakIn2D.java
// https://leetcode.com/problems/find-a-peak-element-ii/
public class PeakIn2D {
    int maxInColumn(int[][] mat, int j) {
        int max = 0;
        for(int i=1; i<mat.length; i++) {
            if(mat[i][j] > mat[max][j]) {
                max = i;
            }
        }
        return max;
    }
    public int[] findPeakGrid(int[][] mat) {
        int n = mat[0].length;
        int l = 0, u = n - 1;
        while(l <= u) {
            int mid = u - (u - l)/2;
            int maxRowInd = maxInColumn(mat, mid);
            int left = mid == 0 ? -1 : mat[maxRowInd][mid - 1];
            int right = mid == n-1 ? -1 : mat[maxRowInd][mid + 1];
            if(mat[maxRowInd][mid] > left && mat[maxRowInd][mid] > right) {
                return new int[]{maxRowInd, mid};
            } else if(mat[maxRowInd][mid] > left) {
                l = mid + 1;
            } else {
                u = mid - 1;
            }
        }
        return new int[]{-1, -1};
    }
}